File: Diagrammatize.Design.Pattern/code/python/Interpreter/Sample/Context.py

from ParseException import ParseException

class Context:
    def __init__(self, text):
        self._tokenizer = text.split()
        self._index = 0
        self._currentToken = None
        self.nextToken()
    
    def nextToken(self):
        if self._index < len(self._tokenizer):
            self._currentToken = self._tokenizer[self._index]
            self._index += 1
        else:
            self._currentToken = None
        return self._currentToken
    
    def currentToken(self):
        return self._currentToken
    
    def skipToken(self, token):
        if token != self._currentToken:
            raise ParseException("Warning: " + token + " is expected, but " + str(self._currentToken) + " is found.")
        self.nextToken()
    
    def currentNumber(self):
        if self._currentToken is None:
            raise ParseException("Warning: No token available")
        
        try:
            return int(self._currentToken)
        except ValueError as e:
            raise ParseException("Warning: " + str(e))

File: Diagrammatize.Design.Pattern/code/cxx/Interpreter/A1/language/Context.cpp
#include "Context.hpp"
#include "ParseException.hpp"
#include <sstream>

namespace {

std::vector<std::string> tokenize(const std::string& text) {
    std::vector<std::string> result;
    std::istringstream iss(text);
    std::string token;

    while (iss >> token) {
        result.push_back(token);
    }
    return result;
}

}   // namespace

Context::Context(const std::string& text) {
    tokens_ = tokenize(text);
    currentIndex_ = 0;
    nextToken();
}

std::string Context::nextToken() {
    if (currentIndex_ < tokens_.size()) {
        currentToken_ = tokens_[currentIndex_];
        currentIndex_++;
    } else {
        currentToken_ = "";
    }
    return currentToken_;
}

const std::string& Context::currentToken() const {
    return currentToken_;
}

void Context::skipToken(const std::string& token) {
    if (token != currentToken_) {
        throw ParseException("Warning: " + token + " is expected, but " + currentToken_ + " is found.");
    }
    nextToken();
}

int Context::currentNumber() {
    if (currentToken_.empty()) {
        throw ParseException("Warning: Empty token");
    }

    try {
        size_t pos;
        int number = std::stoi(currentToken_, &pos);

        // 检查是否整个字符串都被转换了
        if (pos != currentToken_.length()) {
            throw ParseException("Warning: Invalid number format: " + currentToken_);
        }
        return number;
    } catch (const std::invalid_argument& e) {
        throw ParseException("Warning: Invalid argument: " + currentToken_);
    } catch (const std::out_of_range& e) {
        throw ParseException("Warning: Number out of range: " + currentToken_);
    }
}
